src/firstboot/firstboot.c

```diff
@@ -1144,20 +1136,34 @@ static int process_root_account(int rfd) {
 
         } else if (arg_delete_root_password)
                 password = hashed_password = PASSWORD_NONE;
-        else
-                password = hashed_password = PASSWORD_LOCKED_AND_INVALID;
+        else if (!arg_root_password && arg_prompt_root_password) {
+                /* If the user was prompted, but no password was supplied, lock the account. */
+                password = PASSWORD_SEE_SHADOW;
+                hashed_password = PASSWORD_LOCKED_AND_INVALID;
+        } else
+                /* Leave the password as is. */
+                password = hashed_password = NULL;
+
+        /* Don't create/modify passwd and shadow if there's nothing to do. */
+        if (!(password || hashed_password || arg_root_shell)) {
+                log_debug("Initialization of root account was not requested, skipping.");
+                return 0;
+        }
 
         r = write_root_passwd(rfd, pfd, password, arg_root_shell);
         if (r < 0)
                 return log_error_errno(r, "Failed to write /etc/passwd: %m");
 
         log_info("/etc/passwd written.");
 
-        r = write_root_shadow(pfd, hashed_password);
-        if (r < 0)
-                return log_error_errno(r, "Failed to write /etc/shadow: %m");
+        if (hashed_password) {
+                r = write_root_shadow(pfd, hashed_password);
```

Comment:
Actually, I think you need to run this unconditionally in case it's creating `/etc/shadow` but change `write_root_shadow` to allow a `NULL` password like `write_root_passwd` was changed.